### PR TITLE
FSPT-727: Add checkbox question type

### PR DESCRIPTION
**Pull request by MarcUsher**
## 🎫 Ticket
https://mhclgdigital.atlassian.net/browse/FSPT-727

## 📝 Description
Adding the Checkbox question type to our form builder and form runner, making use of a lot of the heavy lifting already done for radios.

- Implements the ability to add a final answer for users if none of the options are appropriate.
- Caps the max checkbox options at 10 (+1 if including a final 'Other' option)
- Makes use of GOV.UK checkbox component's 'exclusive' option for the final 'other' option which will deselect any other checkboxes if selected, and vice versa. Also adds a fallback validator to the form in case JS is disabled
- Export to CSV lists choices as line-separated strings

Still to do in subsequent PRs:
- Update e2e tests
- Add a possible Condition for Checkbox questions

## 🧪 Testing
- Add Checkbox question types with and without the 'other' fallback option
- Edit the questions and see that the options and data is updated correctly
- Test the task/report and check the database and CSV export that the choices are correctly recorded/stored/serialised etc.

### Review Readiness
- [X] PR title and description provide sufficient context for reviewers
- [X] Commits are logical, self-contained, and have clear descriptions

### Performance and security
- ~[ ] No N+1 query problems introduced~
- ~[ ] Any new DB queries include appropriate where clauses based on the user's permissions~

### Testing
- [X] I have tested this change and it meets the acceptance criteria for the ticket
- [X] I need the reviewer(s) to pull and run this change locally
- [X] New (non-developer) functionality has appropriate unit and integration tests
- ~[ ] End-to-end tests have been updated (if applicable)~ _E2E test flow will be updated in a subsequent PR_ 
- [X] Edge cases and error conditions are tested

### CODE REVIEW

File: app/common/templates/common/partials/answers/multiple_choice_from_list.html [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<ul class="govuk-list govuk-list--bullet">
+  {% for choice in answer["choices"] %}
+    <li>
+      <a data-testid="{{ choice['label'] }}-{{ question.text }}">{{ choice["label"] }}</a>
```
Comment (samuelhwilliams) [inline app/common/templates/common/partials/answers/multiple_choice_from_list.html:4]: Why `<a>` elements?

Comment (MarcUsher) [reply app/common/templates/common/partials/answers/multiple_choice_from_list.html:4]: Oh I think because I lazily copied the govuk list from the design system page and it was a list of links and I didn't notice 🤦 I will update, thank you!

File: app/common/collections/types.py [hunk 2]
```diff
@@ -109,6 +109,26 @@ def get_value_for_text_export(self) -> str:
         return self.label
 
 
+class MultipleChoiceFromListAnswer(SubmissionAnswerBaseModel):
+    choices: list[dict[str, str]]
```
Comment (samuelhwilliams) [inline app/common/collections/types.py:113]: Can we make this a TypedDict for better intellisense/checking/autocomplete? It should only be dicts with the keys `key` and `label` right?

File: app/common/collections/forms.py [hunk 3]
```diff
@@ -189,6 +194,24 @@ class _DynamicQuestionForm(DynamicQuestionForm):  # noqa
                     ),
                 ],
             )
+        case QuestionDataType.CHECKBOXES:
+            choices = [(item.key, item.label) for item in question.data_source.items]
+            validators: list[Callable[[Any, Any], None]] = [DataRequired("Select one or more options")]
```
Comment (samuelhwilliams) [inline app/common/collections/forms.py:199]: Can we try using `question.name` instead of `one or more options` here?

File: app/common/forms/validators.py [hunk 4]
```diff
@@ -100,3 +102,31 @@ def __call__(self, form: BaseForm, field: StringField, message: str | None = Non
...
+    def __call__(self, form: BaseForm, field: SelectMultipleField) -> None:
+        if not field.data:
+            return  # Don't validate empty fields - use DataRequired for that
+
+        checkbox_choices = field.data
+        # MyPy expects field.choices to be a dict[str, Any] but in our implementation with wtforms it's a list of tuples
+        form_choices = cast(List[Tuple[str, str]], field.choices)
+        final_option_key, final_option_label, *_ = form_choices[-1]
+        if final_option_key in checkbox_choices and len(checkbox_choices) > 1:
+            message = self.message or f"Select one or more of the options, or select {final_option_label}"
```
Comment (samuelhwilliams) [inline app/common/forms/validators.py:131]: Can we try using `question.name` instead of `one or more of the options` here?

File: tests/unit/deliver_grant_funding/test_forms.py [hunk 5]
```diff
@@ -110,7 +110,7 @@ def test_user_already_platform_admin(app: Flask, factories):
 
 class TestQuestionForm:
     def test_max_data_source_items(self, app):
```
Comment (samuelhwilliams) [inline tests/unit/deliver_grant_funding/test_forms.py:112]: Can we get some of these for checkboxes please 🙏

File: app/common/forms/fields.py [hunk 6]
```diff
@@ -81,3 +81,36 @@ def map_gov_params(self, field, **kwargs):  # type: ignore
             },
         )
         return params
+
+
+class MHCLGCheckboxesInput(MHCLGDividableIterableBase):
```
Comment (samuelhwilliams) [inline app/common/forms/fields.py:86]: Just a couple tests for these (ala the radio input above) cheers 🙇

File: app/developers/data/grants.json [hunk 7]
```diff
@@ -945,6 +1012,19 @@
...
+          "form_id": "615f5183-f056-42c4-acd1-23d25db5f9a3",
+          "hint": "",
+          "id": "e5ef3248-6c32-4554-b4a7-4d286cdb3f6a",
+          "name": "lead contact's favourite GOV.UK component",
+          "order": 7,
+          "presentation_options": {
+            "last_data_source_item_is_distinct_from_others": true
+          },
+          "slug": "what-is-the-lead-contacts-favourite-govuk-component",
+          "text": "What is the lead contact's favourite GOV.UK component?"
```
Comment (samuelhwilliams) [inline app/developers/data/grants.json:1026]: optional nit: 🤏 _most_ of the questions in the seeded grant try to keep a veneer of "realistic monitoring report" to them - would be nice to not bail on that entirely, though we did probably get a little bit loose on some questions before this one so poor precedent set.

Comment (MarcUsher) [reply app/developers/data/grants.json]: Yeah that's fair enough 😄 have updated this and the previous one!
